src/internet_identity/src/storage.rs

```diff
@@ -56,69 +87,115 @@
+
+const DEFAULT_ENTRY_SIZE_V1: u16 = 2048;
+const DEFAULT_ENTRY_SIZE_V3: u16 = 4096;
+
 const EMPTY_SALT: [u8; 32] = [0; 32];
 const GB: u64 = 1 << 30;
-const STABLE_MEMORY_SIZE: u64 = 8 * GB;
+
+/// In practice, II has 32 GB of stable memory available. But we want to keep the default
+/// user range until the stable memory migration is complete. Thus we keep this value for anchor
+/// range checking for the time being.
+const LEGACY_STABLE_MEMORY_SIZE: u64 = 8 * GB;
 /// We reserve last ~10% of the stable memory for later new features.
-const STABLE_MEMORY_RESERVE: u64 = STABLE_MEMORY_SIZE / 10;
+const STABLE_MEMORY_RESERVE: u64 = LEGACY_STABLE_MEMORY_SIZE / 10;
 
 const PERSISTENT_STATE_MAGIC: [u8; 4] = *b"IIPS"; // II Persistent State
 
 /// The maximum number of users this canister can store.
-pub const DEFAULT_RANGE_SIZE: u64 =
-    (STABLE_MEMORY_SIZE - RESERVED_HEADER_BYTES - STABLE_MEMORY_RESERVE)
-        / DEFAULT_ENTRY_SIZE as u64;
+pub const DEFAULT_RANGE_SIZE_V1: u64 =
+    (LEGACY_STABLE_MEMORY_SIZE - RESERVED_HEADER_BYTES_V1 - STABLE_MEMORY_RESERVE)
+        / DEFAULT_ENTRY_SIZE_V1 as u64;
 
 pub type Salt = [u8; 32];
 
 /// Data type responsible for managing user data in stable memory.
-pub struct Storage<T, M> {
+pub struct Storage<M> {
     header: Header,
     memory: M,
-    _marker: PhantomData<T>,
 }
 
 #[repr(packed)]
 struct Header {
     magic: [u8; 3],
     // version 0: invalid
     // version 1: genesis layout, might have persistent state
-    // version 2: genesis layout, must have persistent state
-    // version 3+: invalid
+    // version 2: migration in progress genesis -> post-migration layout, must have persistent state
+    // version 3: post-migration layout
     version: u8,
     num_users: u32,
     id_range_lo: u64,
     id_range_hi: u64,
     entry_size: u16,
     salt: [u8; 32],
+    entry_size_new: u16,       // post-migration entry size
+    new_layout_start: u32,     // all records < this value are still in the old layout
+    migration_batch_size: u32, // batch size for incremental anchor migration
+}
+
+struct RecordMeta {
+    layout: Layout,
+    offset: u64,
+    entry_size: u16,
+}
+
+impl RecordMeta {
+    pub fn layout_v1(record_number: u32, entry_size: u16) -> Self {
+        RecordMeta {
+            layout: Layout::V1,
+            offset: RESERVED_HEADER_BYTES_V1 + record_number as u64 * entry_size as u64,
+            entry_size,
+        }
+    }
+
+    pub fn layout_v3(record_number: u32, entry_size: u16) -> Self {
+        RecordMeta {
+            layout: Layout::V3,
+            offset: RESERVED_HEADER_BYTES_V3 + record_number as u64 * entry_size as u64,
+            entry_size,
+        }
+    }
+
+    pub fn candid_size_limit(&self) -> usize {
+        // u16 is the length of candid before the actual candid starts
+        self.entry_size as usize - std::mem::size_of::<u16>()
+    }
```

Comment:
Can you clarify what this is? does this mean we have `size - 2 bytes` for candid, because we prefix the record with the size of the serialized candid structured and hence here return the actual size available for the serialized candid?

Reply:
Yes, that's it. I'll make the comment clearer.